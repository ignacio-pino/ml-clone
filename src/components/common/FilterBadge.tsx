import { Tag, TagLabel, TagCloseButton } from "@chakra-ui/react";
import { IFilterList } from "../../Interfaces";
import { useLocation, useNavigate, createSearchParams } from "react-router-dom";

export default function FilterBadge({ id, values }: IFilterList) {
  const { search } = useLocation();
  const searchParams = new URLSearchParams(search);

  const navigate = useNavigate();

  const params = Object.fromEntries(searchParams.entries());

  const handleClick = () => {
    delete params[id];
    navigate({
      pathname: "/products",
      search: `${createSearchParams(params)}`,
    });
  };

  return (
    <Tag
      size={"sm"}
      key={id}
      variant="subtle"
      backgroundColor={"purple.400"}
      textColor={"purple.900"}
      w={"fit-content"}
    >
      <TagLabel>{values[0].name}</TagLabel>
      <TagCloseButton
        ml={"0.3rem"}
        onClick={() => handleClick()}
        aria-label="Eliminar filtro"
      />
    </Tag>
  );
}
