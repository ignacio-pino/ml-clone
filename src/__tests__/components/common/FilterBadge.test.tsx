import { render, screen, fireEvent } from "@testing-library/react";
import FilterBadge from "../../../components/common/FilterBadge";
import { MemoryRouter, createSearchParams } from "react-router-dom";

const mockedUseNavigate = vitest.fn();

vitest.mock("react-router-dom", () => ({
  ...require("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("FilterBadge Component", () => {
  const mockFilter = {
    id: "testId",
    name: "testName",
    type: "text",
    values: [{ id: "valuesTestId", name: "valuesTestName" }],
  };

  it("matches snapshot", () => {
    const params = { category: "test" };
    const container = render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <FilterBadge {...mockFilter} />
      </MemoryRouter>
    ).container;

    expect(container).toMatchSnapshot();
  });

  it("renders the filter badge with correct text", () => {
    const params = { category: "test" };
    render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <FilterBadge {...mockFilter} />
      </MemoryRouter>
    );
    const badge = screen.getByText("valuesTestName");
    expect(badge).toBeInTheDocument();
  });

  it("calls the handleClick function when close button is clicked", () => {
    const params = { category: "test", testId: "testId" };
    render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <FilterBadge {...mockFilter} />
      </MemoryRouter>
    );

    const closeButton = screen.getByLabelText(/Eliminar filtro/i);
    fireEvent.click(closeButton);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "category=test",
    });
  });
});
