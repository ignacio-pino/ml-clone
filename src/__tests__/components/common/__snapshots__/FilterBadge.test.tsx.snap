// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterBadge Component > matches snapshot 1`] = `
<div>
  <span
    class="css-12nbfm5"
  >
    <span
      class="css-zvlevn"
    >
      valuesTestName
    </span>
    <button
      aria-label="Eliminar filtro"
      class="css-1t3z1vh"
      type="button"
    >
      <svg
        class="chakra-icon css-1md6f3y"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
          fill="currentColor"
        />
      </svg>
    </button>
  </span>
</div>
`;
